### DISCUSSION TITLE
Can you place the admin panel behind a login form ?

### DISCUSSION THREAD

**Discussion by GeorgeFlorian** [pocketbase/pocketbase#2237] [Q&A]
Hello.

I am looking to create a simple portfolio website, but since I am just a junior React Dev I have no freaking clue on how to set up the whole environment: docker, front-end, back-end, hosting, server and make them into one package that works. I am still working on that.
Funny thing is that I can't find any tutorial on the basics like: `how to put PocketBase on a host`.

My concern is that I don't understand how to place the admin panel behind a login form or if it is even possible.

Let's assume that I've put everything on a single hosting: `https://my-app.com`:
 - `https://my-app.com` is where my landing page is
 - `https://my-app.com:8080/admin` is where I would like to be able to access the admin panel

I don't want everybody to access the admin panel so I've set it up on another port and I have placed it behind a login form.
Now, since this is a fresh app, how does it know who is admin if no users were created ? How can I create an admin user that can access the admin panel ?

I have seen this line of code: `const authData = await pb.admins.authWithPassword('test@example.com', '1234567890');` but do I want my admin credentials in plain code ?

**Comment by ganigeorgiev** [comment]
I'm sorry but I don't understand your questions and what you are trying to do.

> I don't want everybody to access the admin panel so I've set it up on another port and I have placed it behind a login form.

The admin panel is already behind a login form. The `/_/` address at the moment can't be changed.

> I have no freaking clue on how to set up the whole environment: docker, front-end, back-end, hosting, server and make them into one package that works. 

The PocketBase executable is a statically linked binary, meaning that there are no external dependencies and to deploy it you have to only upload the executable (and eventually `pb_public` and `pb_migrations` directories) on your server. For minimal guide you can follow https://pocketbase.io/docs/going-to-production/.

>  Now, since this is a fresh app, how does it know who is admin if no users were created ?

The admin registration is public only until the fist admin account is created. After that the registration form is "locked" and only other admins can add new admins and access the Admin UI.
You can also create the first admin with a migration script. For a JS migration example you can check https://github.com/pocketbase/pocketbase/discussions/1354.

> I have seen this line of code: const authData = await pb.admins.authWithPassword('test@example.com', '1234567890'); but do I want my admin credentials in plain code ?

The above is suitable only in a server-side context. If your app is entirely client-side (eg. SPA), then you shouldn't hardcore any credentials since they will be accessible by the client.

**Reply by GeorgeFlorian** [reply]
> I'm sorry but I don't understand your questions and what you are trying to do.

I did not expect such a fast answer. Thank you.
I want to create a portfolio website where I can use the PocketBase admin panel to upload projects and React to display them.
I want to put the whole project: frontend and backend on the same host.

I am very new at everything that is related to hosting, setting up environment and deploying.

My biggest issue is that I don't understand how do you even build this in one package that will run on a host ?
![image](https://user-images.githubusercontent.com/39895513/230461268-728ec0bf-5997-4730-9db7-546818d4f7b0.png)

Did I mention I am a beginner ? xD

**Reply by ganigeorgiev** [reply]
> RIght now I can't even set up an npm script to start pocketbase and the react app at the same time:

Both `vite` and the `pocketbase` script are long running processes and I don't think you can chain them like that. To avoid blocking you can add `&` to each individual command to run them in the background, but I wouldn't recommend this since you'll loose the easy access to the console output of the commands. It would be better and simpler to just execute them separately in 2 terminal windows/tabs.

> My biggest issue is that I don't understand how do you even build this in one package that will run on a host ?

I'm not sure how to help you with this, since there are multiple ways to do it and depends how you plan to interact with PocketBase - client-side (aka. browser), server-side (eg. node) or both (eg. for SSR).

The easiest setup would be if your frontend app is entirely client-side, aka. single-page-application (SPA) or static prerendered htmls because you can put all static resources in a `pb_public` directory next to the executable and PocketBase will take care to serve its content. This way you'll have to deploy only the PocketBase executable and the `pb_public` dir (_and `pb_migrations` in case you want both local and prod envs to have the same collections configuration_).

**Reply by GeorgeFlorian** [reply]
> The easiest setup would be if your frontend app is entirely client-side, aka. single-page-application (SPA) or static prerendered htmls because you can put all static resources in a `pb_public` directory next to the executable and PocketBase will take care to serve its content. This way you'll have to deploy only the PocketBase executable and the `pb_public` dir (_and `pb_migrations` in case you want both local and prod envs to have the same collections configuration_).

That sounds great.

Since I want to make a portfolio website and it does not have users, under what user do I fetch something from a collection ?
And if it's under an admin user `const authData = await pb.admins.authWithPassword('test@example.com', '1234567890');` how do I protect the admin auth details on an SPA ?
Locally I can set `.env` and add stuff there, but if I put all static resources in the `pb_public` directory, will PocketBase read something like `.env` ?

**Reply by ganigeorgiev** [reply]
>  how do I protect the admin auth details on an SPA ?

You don't have to authenticate as admin if you want to fetch data that would be public accessible.
You can manage the visibility of the collections via the API rules. You can find more info in https://pocketbase.io/docs/api-rules-and-filters/.
